GitCommands/Git/GitRef.cs

```diff
@@ -1,16 +1,28 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Linq;
-using GitCommands.Config;
 using GitUIPluginInterfaces;
 using JetBrains.Annotations;
 
 namespace GitCommands
 {
+    internal enum GitRefType
+    {
+        Other,
+        Head,
+        Remote,
+        Tag,
+        Bisect,
+        BisectGood,
+        BisectBad,
+        Stash
```

Comment:
Not sure if it is this commit, but viewing stash gives the following exception:

![image](https://user-images.githubusercontent.com/6248932/42788934-b9e32a44-8962-11e8-83e3-6779fca3bc51.png)

 	[External Code]	
>	GitExtUtils.dll!GitUI.ThreadHelper.CompletedResult<GitCommands.GitSubmoduleStatus>(System.Threading.Tasks.Task<GitCommands.GitSubmoduleStatus> task) Line 128	C#
 	GitUI.dll!GitUI.FileStatusList.AppendItemSubmoduleStatus(string text, GitCommands.GitItemStatus item) Line 379	C#
 	GitUI.dll!GitUI.FileStatusList.FileStatusListView_DrawItem(object sender, System.Windows.Forms.DrawListViewItemEventArgs e) Line 332	C#
 	[External Code]	
 	GitUI.dll!GitUI.UserControls.ExListView.WndProc(ref System.Windows.Forms.Message m) Line 236	C#
 	[External Code]	
 	GitUI.dll!GitUI.UserControls.ExListView.WndProc(ref System.Windows.Forms.Message m) Line 226	C#
 	[External Code]	
 	GitUI.dll!GitUI.FileStatusList.FileStatusListView_SizeChanged(object sender, System.EventArgs e) Line 1069	C#
 	[External Code]	
 	GitUI.dll!GitUI.UserControls.ExListView.WndProc(ref System.Windows.Forms.Message m) Line 270	C#
 	[External Code]	
 	GitUI.dll!GitUI.UserControls.ExListView.WndProc(ref System.Windows.Forms.Message m) Line 270	C#
 	[External Code]	
 	GitUI.dll!GitUI.FileStatusList.HandleVisibility_NoFilesLabel_FilterComboBox(bool filesPresent) Line 1236	C#
 	GitUI.dll!GitUI.FileStatusList.UpdateFileStatusListView(bool updateCausedByFilter) Line 814	C#
 	GitUI.dll!GitUI.FileStatusList.GitItemStatusesWithParents.set(System.Collections.Generic.IReadOnlyList<(GitCommands.GitRevision revision, System.Collections.Generic.IReadOnlyList<GitCommands.GitItemStatus> statuses)> value) Line 806	C#
 	GitUI.dll!GitUI.FileStatusList.SetDiffs(System.Collections.Generic.IReadOnlyList<GitCommands.GitRevision> revisions) Line 1228	C#
 	GitUI.dll!GitUI.CommandsDialogs.RevisionDiffControl.DisplayDiffTab() Line 142	C#
 	GitUI.dll!GitUI.CommandsDialogs.FormBrowse.FillDiff() Line 1056	C#
 	GitUI.dll!GitUI.CommandsDialogs.FormBrowse..ctor.AnonymousMethod__52_1(object sender, System.EventArgs e) Line 256	C#
 	GitUI.dll!GitUI.RevisionGridControl..ctor.AnonymousMethod__100_0(object _, System.EventArgs e) Line 134	C#
 	[External Code]	
 	GitUI.dll!GitUI.GitUICommands.StartBrowseDialog(System.Windows.Forms.IWin32Window owner, string filter, GitUIPluginInterfaces.ObjectId selectedCommit, bool startWithDashboard) Line 1073	C#
 	GitUI.dll!GitUI.GitUICommands.RunBrowseCommand(System.Collections.Generic.IReadOnlyList<string> args) Line 1555	C#
 	GitUI.dll!GitUI.GitUICommands.RunCommandBasedOnArgument(System.Collections.Generic.IReadOnlyList<string> args, System.Collections.Generic.IReadOnlyDictionary<string, string> arguments) Line 1387	C#
 	GitUI.dll!GitUI.GitUICommands.RunCommand(System.Collections.Generic.IReadOnlyList<string> args) Line 1356	C#
 	GitExtensions.exe!GitExtensions.Program.RunApplication() Line 130	C#
 	GitExtensions.exe!GitExtensions.Program.Main() Line 56	C#

Reply:
This is when selecting a stash by the way...

Reply:
I cannot reproduce this. Looks like you're clicking on a stash in the revision grid to trigger this. Can you describe the stash for me? I've tried a few things here and failed. Were you in a subproject or a superproject?

I'd need the inner stack trace as well. The one you've posted gets the `AggregateException` raised by the `task.Result` call. The real NRE happened asynchronously within the task responsible for populating submodule statuses. That happens in both `GitModule.GetCurrentSubmoduleStatus` and `GitModule.GetSubmoduleStatus`.

R# points out there might be a NRE in the latter, which I'll push a patch for. It might be the fix, but I can't verify until able to reproduce.

Reply:
I get this for private changes in my RunnerUp repo when I first select Working dir and diff tab with one changed submodule and a few changes, then select stash on a another branch. I have not been able to reproduce in Gitextensions repo, will try some more.
 
  | Name | Value | Type
-- | -- | -- | --
  | StackTrace | "   vid GitCommands.GitSubmoduleStatus.CheckSubmoduleStatus(GitModule submodule) i F:\\dev\\gc\\gitextensions_4\\GitCommands\\Git\\GitSubmoduleStatus.cs:rad 41\r\n   vid GitCommands.GitModule.<>c__DisplayClass192_1.<<GetSubmoduleStatus>b__1>d.MoveNext() i F:\\dev\\gc\\gitextensions_4\\GitCommands\\Git\\GitModule.cs:rad 2557\r\n--- Slut på stackspårningen från föregående plats där ett undantag utlöstes ---\r\n   vid System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   vid System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   vid Microsoft.VisualStudio.Threading.JoinableTask.<JoinAsync>d__78.MoveNext()\r\n--- Slut på stackspårningen från föregående plats där ett undantag utlöstes ---\r\n   vid System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   vid System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   vid Microsoft.VisualStudio.Threading.JoinableTask`1.<JoinAsync>d__3.MoveNext()" | string

Reply:
I have a hard time reproducing this outside of the debugger at home. I have seen similar issues before also in normal builds, more in v3-ux, but likely also in master. Then it has been sporadic.
So this is probably not something new in v3-ux, maybe it occurs more often (?) due to optimizations even.

It seems like the problem occur if switching from the repo before submodule calculations are finished. This would explain why I have seen this more often in my daytime life where many recursive levels of submodules are used rather than with GE development (the test setup has a slightly heavier submodule too).

Another occurrence, pointing to the same analysis @drewnoakes made.

  | Name | Value | Type
-- | -- | -- | --
  | StackTrace | "   vid System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)\r\n   vid GitUI.ThreadHelper.CompletedResult[T](Task`1 task) i F:\\dev\\gc\\gitextensions_4\\GitExtUtils\\GitUI\\ThreadHelper.cs:rad 128\r\n   vid GitUI.FileStatusList.AppendItemSubmoduleStatus(String text, GitItemStatus item) i F:\\dev\\gc\\gitextensions_4\\GitUI\\UserControls\\FileStatusList.cs:rad 379\r\n   vid GitUI.FileStatusList.FileStatusListView_DrawItem(Object sender, DrawListViewItemEventArgs e) i F:\\dev\\gc\\gitextensions_4\\GitUI\\UserControls\\FileStatusList.cs:rad 332\r\n   vid System.Windows.Forms.ListView.OnDrawItem(DrawListViewItemEventArgs e)\r\n   vid System.Windows.Forms.ListView.CustomDraw(Message& m)" | string